### ISSUE TITLE
vmstorage cache not expiring until restart

### ISSUE THREAD

**Issue by naveen-vellingiri** [VictoriaMetrics/VictoriaMetrics#7372]
### Is your question request related to a specific component?

vmstorage

### Describe the question in detail

I am using clustered version of victoriametrics(5 vmstorage nodes), The ingestion rate is more than 1M/sec, But recently we experienced a cardinality explosion issue in our stack which was unnoticed and later we found out and blocked those metrics, But due to this the vmstorage memory usage has grown drastically from 100GB to 350GB per storage node (limit is 400GB), Even after blocking those metrics it reduced only by 20GB 320GB/node, Again, a restart of vmstorage node but after that it reduced drastically(to <100GB) , I am assuming that here it's because of cache that vmstorage was using more memory, but ideally since the default cache expiration is set to 30m, the cache should expire but that doesn't happening, Also a point to note we also have a vmalert setup which continuously queries this every 15sec,

Attaching screenshot for ref:
Cache size before restart:
<img width="257" alt="image" src="https://github.com/user-attachments/assets/25aafab7-ee17-4681-8e9f-0a279126a928">

Cache size after restart:
<img width="281" alt="image" src="https://github.com/user-attachments/assets/05169b4f-5e84-446f-a52a-e6ccb6b66dc3">

Even though we can see very minimal decrease in cache size after restart too but look at the vmstorage memory usage/pod after restart:
Before restart:
<img width="556" alt="image" src="https://github.com/user-attachments/assets/c98563af-416b-4339-b314-fe8aa9c6ce35">

After Restart:
<img width="326" alt="image" src="https://github.com/user-attachments/assets/40a8a678-2c0c-4abe-90ab-58366c037d9c">

### Troubleshooting docs

- [X] General - https://docs.victoriametrics.com/troubleshooting/
- [X] vmagent - https://docs.victoriametrics.com/vmagent/#troubleshooting
- [X] vmalert - https://docs.victoriametrics.com/vmalert/#troubleshooting

**Comment by jiekun** [comment]
Hi. 

The cache could work in different `mode`: `split` mode and `whole` mode.

For `split` mode, cache has chance to expire, swap and memory could be released. See: https://github.com/VictoriaMetrics/VictoriaMetrics/blob/d845edc24b0eae5d24466ce1d30d09c3b6fb0b37/lib/workingsetcache/cache.go#L135
For `whole` mode, cache use all space and never release them. See: https://github.com/VictoriaMetrics/VictoriaMetrics/blob/d845edc24b0eae5d24466ce1d30d09c3b6fb0b37/lib/workingsetcache/cache.go#L218

The mode switching happens when cache is using more than 50% of space (of `allowPercent`). If you set `allowedPercent` to 40%, and somehow experience a churn rate spike, the memory usage exceed 50% of it cap, it will transit to `whole` mode and use 100% of space

Please see also: https://github.com/VictoriaMetrics/VictoriaMetrics/issues/6701#issuecomment-2257439138

**Comment by naveen-vellingiri** [comment]
@jiekun  Acknowledged, 

Also, If a mode switch happens, then the cache occupied by the whole mode will be considered in the `vm_cache_size_bytes` metric?


Because in our case we don't see a drastic decrease in cache usage after restart, But in memory yes. 

Attaching screenshot for ref:
<img width="1120" alt="Screenshot 2024-10-28 at 8 49 47 PM" src="https://github.com/user-attachments/assets/8ea3aa9a-1b41-4f72-bcf6-957f622c1dc8">
In the above image point A is at the time of high churn rate, point b is after resolution , point c is after restart

**Comment by jiekun** [comment]
Hi! could you also share the cache usage at that time?

**Comment by naveen-vellingiri** [comment]
At the time of high churn rate:
<img width="1143" alt="image" src="https://github.com/user-attachments/assets/174c6923-e4e6-4d45-a0ad-d8423196c367">

After resolution:
<img width="1142" alt="image" src="https://github.com/user-attachments/assets/4a6e3cc7-a5cf-4c77-a869-bb4f2de60df6">

After restart:
<img width="1141" alt="image" src="https://github.com/user-attachments/assets/8fca107d-cfc6-4599-bbd1-1b1fc699d231">

Query used:
<img width="1160" alt="image" src="https://github.com/user-attachments/assets/9c8b6dae-4750-4179-b9b1-e3d617a89c7d">

**Comment by jiekun** [comment]
Hi. Thanks for the info. I'd like to ask for the cache usage of different caches. The following MetricsQL should work for you:

```
max(
    vm_cache_size_bytes{job=~"$job", instance=~"$instance"} 
    /
    vm_cache_size_max_bytes{job=~"$job", instance=~"$instance"}
) by(type)
```
Here's an example:

<details>
![image](https://github.com/user-attachments/assets/a49dd60c-b9e3-418d-829a-0cc8859bac58)
</details>

**Comment by naveen-vellingiri** [comment]
Attaching the different cache usage screenshots:
At high churn rate:
<img width="1157" alt="image" src="https://github.com/user-attachments/assets/8edc736d-c460-411f-b902-4955aecb66ab">

After resolution of high churn rate:
<img width="1159" alt="image" src="https://github.com/user-attachments/assets/5c438a2a-7fb1-4587-830b-50fe7e2fc80b">

After restart of vmstorage:
<img width="1157" alt="image" src="https://github.com/user-attachments/assets/a009365b-fa0d-4d9e-9237-614631170112">

Usage pattern over last 10 days:
<img width="1158" alt="image" src="https://github.com/user-attachments/assets/73c8f5a5-169c-407f-bfa9-86d994f11236">


query used:
```(max(
    vm_cache_size_bytes{job=~"$job",kubernetes_pod_name=~"${pods_storage}"} 
    /
    vm_cache_size_max_bytes{job=~"$job", kubernetes_pod_name=~"${pods_storage}"}
) by(type))*100

**Comment by jiekun** [comment]
Thank you! The cache usage and the cache size seem to be mismatched. To me, the cache usage before/after the restart looks **stable**, while the cache size dropped significantly.

I'd like to suggest focusing on one type of cache first and seeing why there is a difference. For example:
1. The cache size of `indexdb/dataBlocks` after the restart is about **19.3 GiB**. Could you check the average size after the resolution of the high churn rate and before the restart?
2. If the cache usage of `indexdb/dataBlocks` is stable after the resolution, could you check if `vm_cache_size_max_bytes` of this cache changed?

**Comment by naveen-vellingiri** [comment]
Thanks @jiekun  for your response, 
Also I anlayzed the cache usage, it seems to be for this type `indexdb/dataBlocks` the usage drastically decreased after resolution of high churn rate: from 480 to 5gb
<img width="873" alt="image" src="https://github.com/user-attachments/assets/99de19b6-1a45-4785-b83b-cb66af2d0f1e">

But this decrease in cache usage is not reflected in the memory usage of vmstorage pods, after resolution it just decreased from 1.98tb to 1.75tb
<img width="866" alt="image" src="https://github.com/user-attachments/assets/fca4d481-9a5d-4199-b46e-b25230ec1e11">

Also, the `storage/metricname` is the cache type where we see reduction in cache usage after restart rest everything looks stable(from resolution of high churn rate to after restart of pods)

**Comment by jiekun** [comment]
I'm checking these screens, and they looks different. Is that correct?

<details>

![image](https://github.com/user-attachments/assets/3214dcac-b02f-4ba2-a79f-c1733053c463)

![image](https://github.com/user-attachments/assets/d8ead40f-8c0f-4684-a7ce-99b181556a50)

</details>

I want to find a cache whose trend is similar to the 1st screen.

Your original question is, why some cache space is not released after the resolution of high churn rate, but drops after restart. So:
1. first of all, I'd like to look for a certain type of cache that match this trend.
2. the second step is to check why the cache usage looks like the same after resolution and restart, according to your comment [here](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/7372#issuecomment-2443874987).

**Comment by naveen-vellingiri** [comment]
@jiekun yes they are different , because the first screen is stacked chart whereas the second is not.
that's why the difference is there, 
Actually only this cache type is where i see a major dip after restart
<img width="1155" alt="image" src="https://github.com/user-attachments/assets/872de9fc-89fe-4c35-b350-539a8a570e21">

Coming to the original question: 
`why some cache space is not released after the resolution of high churn rate, but drops after restart`

But the point here is, even though the cache memory is released (as per the metric) after resolution of high churn rate (https://github.com/VictoriaMetrics/VictoriaMetrics/issues/7372#issuecomment-2446018720)
, this is not reflected in vmstorage's memory usage until a restart is happend.

**Comment by jiekun** [comment]
Hi. Could you share the whole VictoriaMetrics Cluster dashboard in the time range you mentioned above via the "share" feature of Grafana?

It's hard for me to align the data you mentioned, and I'd like to check them at once.

![image](https://github.com/user-attachments/assets/195a2133-9057-4f54-9ff1-445796298542)

**Comment by jiekun** [comment]
Hello! @naveen-vellingiri Do you still need help on this issue?
